tests/conftest.py

```diff
@@ -1,9 +1,185 @@
 """Fixtures common across tests."""
+import os
+import pickle
+import traceback
+import warnings
+from typing import Callable, List
 
+import gym
 import pytest
 import torch
+from filelock import FileLock
+from stable_baselines3 import PPO
+from stable_baselines3.common.policies import BasePolicy
+from stable_baselines3.common.vec_env import DummyVecEnv, VecNormalize
 
+from imitation.data import rollout
+from imitation.data.types import TrajectoryWithRew
+from imitation.data.wrappers import RolloutInfoWrapper
+from imitation.policies.base import FeedForward32Policy, NormalizeFeaturesExtractor
 from imitation.util import logger
+from imitation.util.networks import RunningNorm
+
+CARTPOLE_ENV_NAME = "CartPole-v1"
+
+
+def load_or_train_ppo(
+    cache_path: str,
+    training_function: Callable[[gym.Env], PPO],
+    venv,
+) -> PPO:
+    os.makedirs(os.path.dirname(cache_path), exist_ok=True)
+    with FileLock(cache_path + ".lock"):
```

Comment:
Just to note this lock is a bit trigger-happy as it'll stop multiple processes from *reading* even if it already exists. But that's probably fine, loading should be quick relative to the rest of the tests, so I don't think we need to change it but just something to watch out for if we do ever see test slowdowns.

Reply:
You mean we should wrap the FileLock only around training and not loading? That would unfortunately make things worse in comparison to no locks at all. Assume the following scenario with workers W1 and W2.
- W1 hits `load_or_train_ppo` first, loading fails, it acquires the lock and then starts training.
- W2 hits `load_or_train_ppo`, loading fails, it tries to acquire the lock but has to wait for W1
- W1 finishes training, releases the lock
- W2 gets the lock and starts training

So you see then we still have double training but now it's not even parallel any more. We need to wrap the FileLock around loading *and* training.

Parallel computing with shared state is a mess ...

Reply:
Yeah, I agree that locking just the training wouldn't work. 

I think there is something more sophisticated one could do, but do not think it is currently worth the added complexity given FileLock does not support this. But, the rough idea I had in mind was:
   1. Obtain a read-only lock. This can be obtained if there are other read-only holders, but *not* if there is a write holder.
   2. Attempt to load the policy.
   3. If loading succeeds, return and release lock.
   4. If loading fails, release read-only lock, then attempt to obtain write lock. This will block until all read-only lock holders relinquish the lock.
   5. Attempt to load policy. (This is needed in case we got the write lock after another process completed training.) Otherwise, train policy, and save it. 
   6. Return and release write lock.

Reply:
I see. That seems to be beyond the scope of the filelock library :-)